### PR TITLE
added new lint `owned_to_owned`

### PR DESCRIPTION
**Pull request by anall**
Adding new lint `owned_to_owned`

Creating draft PR to have this looked over.
I think this takes all advice I received into account.

I did have to update the `redundant_clone` test to ignore this lint -- I felt that was the safest action.

closes: #6715
changelog: added new lint `implicit_clone`

### CODE REVIEW

File: clippy_lints/src/methods/mod.rs
Comment (camsteffen) [inline clippy_lints/src/methods/mod.rs:4204]: You may change `args` (above) to `[arg]`

**Comment by anall** [comment]
Should I move my implementation to `methods/owned_to_owned.rs` to give less work for whoever takes up `methods` as part of #6680?

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (camsteffen) [inline clippy_lints/src/methods/owned_to_owned.rs:20]: I believe the "expected type" check is redundant since you are verifying the method.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (camsteffen) [inline clippy_lints/src/methods/owned_to_owned.rs:35]: "Using `to_vec` on a `Vec`" would be more clear.

```suggestion
                    &format!("using `{}` on a `{}`", method_path.ident.name, ty_name),
```

where `ty_name` is `cx.tcx.item_name(input_type.ty_adt_def().unwrap().did)`

Also (very picky nit) please remember spaces after commas inside `if_chain!`.

**Review by camsteffen — CHANGES_REQUESTED** [review]
This is looking great! I know this PR is a draft but it seems nearly done to me 😄.

> I did have to update the `redundant_clone` test to ignore this lint -- I felt that was the safest action.

Makes sense. And I think it's okay if those lints both fire on the same code.

You need tests for calling on a reference to a `Vec` (or other). You will probably need `input_type.peel_refs()` before calling `same_type`.

I'm sorry to bikeshed even more but I'm having second thoughts on the lint name. How about `implicit_clone`? It seems to me that the fact that we are dealing with owned types is not actually very relevant. The problem being linted is that the code is using a round-about way (deref + to_owned) to clone something rather than cloning explicitly. This lint might apply to a `to_something` method that does *not* produce an "owned" type. Or, this lint could apply to usage of `from` or `into` to clone something. I think this name is more self-explanatory too.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [reply clippy_lints/src/methods/owned_to_owned.rs:11]: I put `lint` here as most other (looks like all but `filter_map_identity`) lints in other files seem to be using `lint`, but happy to make that change. And happy to make the other changes.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [reply clippy_lints/src/methods/owned_to_owned.rs:20]: For the current version here, I believe you are right but further changes in prep of using diagnostic items instead (including the ones I'm trying to get added in https://github.com/rust-lang/rust/pull/82128) checks _either_ the trait of the method (for `to_owned`) or the return type (for `to_vec`, …)

I can commit/push a new checkpoint if you'd like.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [reply clippy_lints/src/methods/owned_to_owned.rs:35]: I was actually looking for how to do that. Done.

Re the nit: Fixed and I will try to be more aware of the missing spaces.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [reply clippy_lints/src/methods/owned_to_owned.rs:20]: I believe this is no longer redundant after some cleanup refactoring + moving to being able to use diagnostic items (including the ones I'm trying to have added in https://github.com/rust-lang/rust/pull/82128)

I haven't committed/pushed that change yet, but I check if the method belongs to a trait_path _or_ the type.

I'll be pushing my new changes once I've gone through everything else.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [reply clippy_lints/src/methods/owned_to_owned.rs:22]: Nothing seems to prevent `to_owned()` from being called on a custom type that implements `Copy` (like the `Kitten` type in my test), do we want this lint to fire on those cases?

Without this check, this will suggest replacing `to_owned` with `clone`, and then `clone_on_copy` will suggest dropping the `.clone()` entirely.

**Comment by anall** [comment]
> You need tests for calling on a reference to a `Vec` (or other). You will probably need `input_type.peel_refs()` before calling `same_type`.

Wouldn't that end up suggesting that even calling `to_owned` on a `&Vec` should be turned into `clone()`? Not objecting, but that seems to be going against the original issue.

I did add a few (not-yet-committed) negative tests
 
> I'm sorry to bikeshed even more but I'm having second thoughts on the lint name. How about `implicit_clone`? It seems to me that the fact that we are dealing with owned types is not actually very relevant. The problem being linted is that the code is using a round-about way (deref + to_owned) to clone something rather than cloning explicitly. This lint might apply to a `to_something` method that does _not_ produce an "owned" type. Or, this lint could apply to usage of `from` or `into` to clone something. I think this name is more self-explanatory too.

That seems quite reasonable -- I'm not sure how GitHub handles tracking changes across renames, so will hold off on the rename until at least the next review pass.

(I know git itself handles that sub-optimally unless you explicitly tell git to follow renames)

File: clippy_lints/src/methods/mod.rs
Comment (giraffate) [inline clippy_lints/src/methods/mod.rs:1540]: nits
```suggestion
    "using `to_owned` on already owned type"
```

File: clippy_lints/src/methods/mod.rs
Comment (giraffate) [inline clippy_lints/src/methods/mod.rs:1521]: nits
```suggestion
    /// to why we are calling `to_vec` on something that is already a `Vec`
```

**Comment by anall** [comment]
I suggested replacing `string_to_string` with this lint, but I'd feel better if that could be in a separate pull request.

I'm also happy waiting for those diagnostic items and doing that change before merging this, or am fine with this being merged before and then changing to the diagnostic items as soon as they are available. Whichever you prefer.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [reply clippy_lints/src/methods/owned_to_owned.rs]: I should clarify: when I say "to be able to use diagnostic items" I mean hopefully easily switch the function from taking paths to the function taking diagnostic items -- not a mix of both.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (camsteffen) [reply clippy_lints/src/methods/owned_to_owned.rs]: Oh I see. It is *technically* possible but I think this situation is unusual enough that we shouldn't have a special case for it. The purpose of having a borrowed variant of a type is to avoid expensive copies. But if a type implements `Copy`, then it shouldn't be expensive to copy. In any case, I think the lint should just fire normally because `clone` is still better than `to_owned`.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (camsteffen) [reply clippy_lints/src/methods/owned_to_owned.rs]: > checks _either_ the trait of the method (for `to_owned`) or the return type (for `to_vec`, …)

I think it would be better to just check the method for every case. If you only check the return type, then you might have a false positive if someone rolls their own trait with `to_vec` with different semantics (even though that would be a little odd). If you just check the method, there should be no such risk.

**Comment by camsteffen** [comment]
> Wouldn't that end up suggesting that even calling `to_owned` on a `&Vec` should be turned into `clone()`?

Yes. I don't know why that would be any less valid. Both `to_owned` and `clone` accept a reference anyways.

> I know it's been suggested (and I agree) that replacing `string_to_string` with this lint is possible, but I'd feel better if that could be in a separate pull request.

A separate PR for that is a good idea. Getting the diagnostic items in this PR is also good.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [reply clippy_lints/src/methods/owned_to_owned.rs]: I guess by "check the method" while using diagnostic types, you mean "check the type/trait the method is on as well as checking the diagnostic item of return value of the type itself"?

Also, I just noticed `os_string.to_os_string()` is actually a method on `OsStr` so I will need to make sure that has a diagnostic item as well.

Comment (camsteffen) [reply clippy_lints/src/methods/owned_to_owned.rs]: "check the method":
1) the method name is "to_vec"
2) the method is associated with type X
3) type X is diagnostic item "slice"

Not sure how to make the "associated" link but maybe `tcx.associated_item` and/or `tcx.impl_of_method`. We might need a util method to make this easier like `is_diagnostic_assoc_fn(def_id, sym::slice, sym::to_vec)`.

File: clippy_lints/src/methods/owned_to_owned.rs
Comment (anall) [inline clippy_lints/src/methods/owned_to_owned.rs:27]: This doesn't seem to work on struct impls:

The ones on struct impls don't work and give me:
```
got associated DefId(1:4405 ~ std[e437]::path::{impl#61}) for DefId(1:4413 ~ std[e437]::path::{impl#61}::to_path_buf) // kind : "std::path::Path"
 --- Some(DefId(1:4403 ~ std[e437]::path::Path)) false
```

While the ones on traits work and get me:
```
got associated DefId(5:424 ~ alloc[88c7]::borrow::ToOwned) for DefId(5:426 ~ alloc[88c7]::borrow::ToOwned::to_owned) // kind : "std::borrow::ToOwned"
 --- Some(DefId(5:424 ~ alloc[88c7]::borrow::ToOwned)) true
```

I need some way to get from the impl def_id to the struct def_id?

File: clippy_lints/src/methods/mod.rs
Comment (camsteffen) [inline clippy_lints/src/methods/mod.rs:1540]:
```suggestion
    "implicitly cloning a value by invoking a function on its dereferenced type"
```

File: clippy_lints/src/methods/implicit_clone.rs
Comment (camsteffen) [inline clippy_lints/src/methods/implicit_clone.rs:30]: Hmm I'm afraid people will read this error and say, "Yeah, so?" 😄  Suggestion:

```text
implicitly cloning a `Vec` by calling `to_vec` on its dereferenced type
```

**Review by camsteffen — CHANGES_REQUESTED** [review]
A few small things and then we'll ship it!

File: clippy_lints/src/methods/implicit_clone.rs
Comment (anall) [reply clippy_lints/src/methods/implicit_clone.rs:26]: Actually, I think it might be. The idea behind this was to prevent something like this from triggering the lint:

```
struct NotAVec {…};

impl NotAVec {
  fn to_vec(self) -> NotAVec { … }
}
```

But now that I am checking where `to_vec` is implemented (is it on slice?), this to_vec won't work because it is on `NotAVec` so there's no need to check the type.

File: clippy_utils/src/lib.rs
Comment (anall) [reply clippy_utils/src/lib.rs:243]: I assumed it would find wide use.

Renamed.

File: clippy_lints/src/methods/implicit_clone.rs
Comment (camsteffen) [reply clippy_lints/src/methods/implicit_clone.rs:26]: Right. So long as you keep the `same_type` check right above.

File: clippy_lints/src/methods/implicit_clone.rs
Comment (anall) [reply clippy_lints/src/methods/implicit_clone.rs:30]: That may be slightly wrong wording in the case of
```
let vec = Vec::new();
let _ = vec.to_vec();
```
but probably not too much of an issue

**Comment by camsteffen** [comment]
Very nice. Please squash into fewer commits and I'll merge.

**Comment by camsteffen** [comment]
Oops I forgot one thing. Please change from `style` to `pedantic` and run `cargo dev update_lints`.

**Comment by camsteffen** [comment]
Thanks! @bors r+

**Pull request merged**
